libnmstate/state.py

```diff
@@ -444,21 +466,38 @@ def merge_routes(self, other_state):
           - Self iface explicitly specified (regardless of routes)
         - Self absent routes overwrite other routes state.
           - Support wildcard for matching the absent routes.
+        - Other routes next hop to DHCP/Autoconf interface will be discarded.
         """
         other_routes = set()
         for ifname, routes in other_state.config_iface_routes.items():
             if (
                 ifname in self.config_iface_routes
                 or self._is_interface_routable(ifname, routes)
             ):
-                other_routes |= set(routes)
+                for route in routes:
+                    if not self._route_is_next_hop_to_auto_iface(
+                        route.to_dict()
+                    ):
+                        other_routes.add(route)
 
         self_routes = {
             route
             for routes in self.config_iface_routes.values()
             for route in routes
             if not route.absent
+            and not self._route_is_next_hop_to_auto_iface(route.to_dict())
         }
+        for ifname, routes in self.config_iface_routes.items():
+            for route in routes:
+                if not route.absent and self._route_is_next_hop_to_auto_iface(
+                    route.to_dict()
+                ):
+                    logging.warning(
+                        "Discarding static routes to DHCP/Autoconf "
+                        "interface {} as nmstate does not support it "
+                        "yet".format(ifname)
```

Comment:
Please use `logging.warning("...%s...", ifname)`. This avoids the string format in case the message is not actually logged.